### ISSUE TITLE
ramon json interface

### ISSUE THREAD

**Issue by alexbaden** [neurodata/ndstore#237]
Currently, the following work on the ab-microns branch:

- GET RAMON objects as JSON: `/nd/sd/<<token>>/<<channel>>/<<id1,id2,...>>/json/`

At the moment, this returns a dictionary object as a JSON string, where each key is an ID and each value is the serialized dictionary representation of the RAMON object. Using new RAMON, this doesn't exactly match the existing HDF5 interface, but the changes are minimal (fields are named and organized slightly differently, but the representation is more compact).

However, the ID of the RAMON object is now included in this dictionary. So delivering the value as a key is redundant. 

- POST RAMON objects as JSON: `/nd/sd/<<token>>/<<channel>>/<<id1>>/json/update (opt)/`

This expects a dictionary object where the key is the ID, and the value is a dictionary that represents the RAMON object. 

**Suggestion:** Change GET and POST to both work on arrays of dictionaries, not dictionaries of dictionaries. Essentially, this is flattening everything from:
```
{"2": {"syn_weight": 0.0, "ann_type": 2, "syn_type": 0, "ann_status": 0, "ann_confidence": 0.8, "ann_id": 2, "syn_centroid": "[]", "ann_author": "unspecified"}}
```
to
```
{"syn_weight": 0.0, "ann_type": 2, "syn_type": 0, "ann_status": 0, "ann_confidence": 0.8, "ann_id": 2, "syn_centroid": "[]", "ann_author": "unspecified"}
```
 
Example:

**Get Single RAMON Object**
GET: `http://localhost:8000/nd/sd/testtoken/testchannel/1/json/`
RETURN: 
```[{"syn_weight": 0.0, "ann_type": 2, "syn_type": 0, "ann_status": 0, "ann_confidence": 0.8, "ann_id": 1, "syn_centroid": "[]", "ann_author": "unspecified"}]```

**Get Multiple RAMON Objects**
GET: `http://localhost:8000/nd/sd/testtoken/testchannel/1,2/json/`
RETURN: 
```[{"syn_weight": 0.0, "ann_type": 2, "syn_type": 0, "ann_status": 0, "ann_confidence": 0.8, "ann_id": 1, "syn_centroid": "[]", "ann_author": "unspecified"},{"syn_weight": 0.0, "ann_type": 2, "syn_type": 0, "ann_status": 0, "ann_confidence": 0.8, "ann_id": 2, "syn_centroid": "[]", "ann_author": "unspecified"}]```

**New Annotation with User Specified ID**
POST: `http://localhost:8000/nd/sd/testtoken/testchannel/2/json/`
DATA: 
```{"syn_weight": 0.0, "ann_type": 2, "syn_type": 0, "ann_status": 0, "ann_confidence": 0.8, "ann_id": 2, "syn_centroid": "[]", "ann_author": "unspecified"}```

**Update Annotation**
POST: `http://localhost:8000/nd/sd/testtoken/testchannel/2/json/update/`
```{"syn_weight": 0.0, "ann_type": 2, "syn_type": 0, "ann_status": 0, "ann_confidence": 0.8, "ann_id": 2, "syn_centroid": "[]", "ann_author": "unspecified"}```
_Note: _ If you don't include `update` in the URL and the RAMON object already exists, the POST fails. 

**New Annotation w/ Next ID**
POST: `http://localhost:8000/nd/sd/testtoken/testchannel/0/json/`
```{"syn_weight": 0.0, "ann_type": 2, "syn_type": 0, "ann_status": 0, "ann_confidence": 0.8, "syn_centroid": "[]", "ann_author": "unspecified"}```
_Note:_ Either omitting `ann_id` or setting it to 0 are both fine. 

We could extend all of the above to support batch annotations, as well. 

There's also the question of whether this should live in a different URL space ( `/nd/ramon/...`?) since it is a metadata only interface, and doesn't upload any paint. We could centralize all JSON queries under `/ramon/` while maintaining backwards compatibility for CAJAL and the HDF5 RAMON interfaces. 

@kunallillaney @randalburns @j6k4m8 @jovo @willgray

**Comment by j6k4m8** [comment]
What I like about having the multi-layers is that we can invariably lookup into the JSON using the same syntax, no matter if we're requesting one or multiple. But this is definitely something I'd be behind if we implemented it as per above!

**Comment by alexbaden** [comment]
I think you could do the same with the above. One object would be a length 1 array, two objects would be length two, etc. 

The problem I see with the multi-layered schema is on posting: it's unclear what the keys should be. And, whenever I design a GET/POST interface, I make sure a user could do a GET request and immediately turn around POSTing it back, without changing fields.

**Comment by willgray** [comment]
TL;DR:  This is totally cool and I imagine we can work with this in whatever form we ultimately choose.  

I lean toward the nested schema because of indexing specific IDs, but would be pretty quick to dynamically create an index...so not sure.  Would you guarantee to return objects in order requested?  Or sequential order or just random?  If order requested, in some sense the user should already have an ordering, which is a vote for the flattened structure.

+1 to "We could centralize all JSON queries under /ramon/ while maintaining backwards compatibility for CAJAL and the HDF5 RAMON interfaces."  I think it clearly makes sense to separate these functions.  

Is interesting to ponder which services are still HDF5 only (v. JSON/blosc) - not sure if  ndio really needs to deal with hdf5 endpoints going forward?  

What about delete and reserve_ids?  Those and batch are the rest of the things we might commonly need that are ramon oriented.

**Comment by randalburns** [comment]
So, I'm of two minds about this. It certainly makes more sense and the data structures are kind of better.  But, there are a couple of issues.  When transacting mutliple objects, the data becomes not self describing, so you need to reference the ids from the URL to the array.  Don't love that.  Jordan's example query doesn't work, because it refers to positional arguments, not objects ids.  So here are the choices:
  (1) everything is dict of dict.  Lookup by id works.  This is awkward for single objects.  NB this is what msyqldb does when returning a single object you still have to take the ...[0]
  (2) array of dicts and we inline the id inside the dict.  This is self describing, but you can no longer look up items by id.  You have to scan them.
  (3) array of dicts, no inline id.  User has to reference/index objects by the url.  I don't like this because the return data is not self describing.

Seems like all choices are imperfect.  I don't have strong feelings.  But, I chose #1 way back ago for HDF5 for the reason stated.
